src/machine/shared_library.rs

```diff
@@ -0,0 +1,269 @@
+                Ok(input) => {
+                    let string = input.to_string();
+                    let query_state = machine.run_query_iter(string);
+                    Box::into_raw(Box::new(query_state))
+                }
+                Err(err) => {
+                    eprintln!("Query input wasn't valid UTF-8: {err:?}");
+                    std::ptr::null_mut()
+                }
+            }
+        }
+
+        /// Cleans up the [`QueryState`] in the associated Scryer [`Machine`].
+        ///
+        /// # Safety
+        ///
+        /// * `query_state` must be a valid mutable pointer to a [`QueryState`] created by [`run_query_iter`]
+        /// that has not yet been freed by [`query_state_free`].
+        /// * There can be only one [`QueryState`] per [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        #[export_name = "scryer_query_state_free"]
+        pub unsafe extern "C" fn query_state_free(query_state: *mut QueryState) {
+            unsafe {
+                drop(Box::from_raw(query_state));
+            }
+        }
+
+        /// Returns a NULL POINTER if no addition iterations, else returns a
+        /// JSON encoded UTF-8 string with one iteration of results from a Scryer Prolog query.
+        ///
+        /// See documentation for known limitations (e.g., concrete goals only).
+        ///
+        /// # Safety
+        ///
+        /// * `query_state` must be a valid mutable pointer to a [`QueryState`] created by [`run_query_iter`]
+        /// that has not yet been freed by [`query_state_free`].
+        /// * There can be only one [`QueryState`] per [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        #[export_name = "scryer_run_query_next"]
+        pub extern "C" fn run_query_next(query_state: &mut QueryState) -> *mut c_char {
+            match query_state.next() {
+                None => std::ptr::null_mut(),
+                Some(Ok(query_resolution_line)) => {
+                    let v = QueryResolution::from(vec![query_resolution_line]).to_string();
+
+                    let obj = serde_json::from_str::<serde_json::Value>(&v).expect("Bad JSON");
+
+                    let output_string = json!({
+                        "status": "ok",
+                        "result": obj
+                    })
+                    .to_string();
+
+                    CString::new(output_string.to_string()).unwrap().into_raw()
+                }
+                Some(Err(err)) => {
+                    eprintln!("Query input wasn't valid UTF-8: {err:?}");
+                    std::ptr::null_mut()
+                }
+            }
```

Comment:
This error should definitively be returned! For this one I suggest just going the webdev way with something like:

```rust
json!({
    "status": "error",
    // All errors should implement Display and therefore have
    // access to the to_string() method.
    "error": err.to_string(),  
})
```